cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -345,6 +362,139 @@ public ApplicationWithPrograms updateApp(ApplicationId appId, AppRequest appRequ
+   * to its config.
+   */
+  private ApplicationUpdateDetail updateApplicationInternal(ApplicationId applicationId,
+                                                             NamespaceId namespaceId,
+                                                             @Nullable String appName,
+                                                             @Nullable String appVersion,
+                                                             @Nullable String currentConfigStr,
+                                                             ProgramTerminator programTerminator,
+                                                             ArtifactDetail artifactDetail,
+                                                             List<ApplicationConfigUpdateAction> updateActions,
+                                                             @Nullable KerberosPrincipalId ownerPrincipal,
+                                                             boolean updateSchedules,
+                                                             String userAction) throws Exception {
+    ApplicationClass appClass = Iterables.getFirst(artifactDetail.getMeta().getClasses().getApps(), null);
+    if (appClass == null) {
+      // This should never happen.
+      throw new IllegalStateException(String.format("No application class found in artifact '%s' in namespace '%s'.",
+                                         artifactDetail.getDescriptor().getArtifactId(), namespaceId));
+    }
+    io.cdap.cdap.proto.id.ArtifactId artifactId =
+      Artifacts.toProtoArtifactId(namespaceId, artifactDetail.getDescriptor().getArtifactId());
+    EntityImpersonator classLoaderImpersonator = new EntityImpersonator(artifactId, this.impersonator);
+    ClassLoader artifactClassLoader =
+      artifactRepository.createArtifactClassLoader(artifactDetail.getDescriptor().getLocation(),
+                                                   classLoaderImpersonator);
+
+    String updatedAppConfig = "";
+    DefaultApplicationUpdateContext updateContext =
+      new DefaultApplicationUpdateContext(namespaceId, applicationId, artifactDetail.getDescriptor().getArtifactId(),
+                                          artifactRepository, currentConfigStr, updateActions);
+
+    // Run config update logic for the application to generate updated config.
+    Object appMain = artifactClassLoader.loadClass(appClass.getClassName()).newInstance();
+    if (!(appMain instanceof Application)) {
+      throw new IllegalStateException(
+        String.format("Application main class is of invalid type: %s",
+                      appMain.getClass().getName()));
+    }
+    Application app = (Application) appMain;
+    Type configType = Artifacts.getConfigType(app.getClass());
+    if (!app.isUpdateSupported()) {
+      String status = String.format("%s failed.", userAction);
+      String errorMessage = String.format("application of type %s does not support update.",
```

Comment:
the type is a java classname that will likely just confused people. You can just say that the application does not support updates.